src/hotspot/share/opto/parse1.cpp

```diff
@@ -1889,19 +1888,37 @@ void Parse::merge_common(Parse::Block* target, int pnum) {
             } else if (!check_elide_phi || !target->can_elide_SEL_phi(j)) {
               phi = ensure_phi(j, nophi);
 
-              // Part-1: only check 'm'. ensure_phi has replaced m with phi.
+              // C2 PEA merges basic blocks incrementally. We have to intercept the phi creation
+              // for SFPT/map.
               if (DoPartialEscapeAnalysis && phi != nullptr) {
-                VirtualState* vs;
-
-                if ((vs = as.as_virtual(phi)) != nullptr) { // m is Virtual
-                  ObjID id = as.is_alias(phi);
-                  ObjectState* pred_os;
+                PartialEscapeAnalysis* pea = PEA();
+                ObjID obj1 = pea->is_alias(m);
+                ObjID obj2 = pea->is_alias(n);
+
+                if (as.contains(obj1)) { // m points to an object that as is tracking.
+                  ObjectState* os1 = as.get_object_state(obj1);
+                  ObjectState* os2 = pred_as.contains(obj2) ? pred_as.get_object_state(obj2) : nullptr;
+                  bool need_update = false;
+
+                  // obj1 != obj2 if n points to something else. It could be the other object, null or a ConP.
+                  // we don't update it here because PEA doesn't create phi in this case.
+                  if (obj1 == obj2 && os2 != nullptr) { // n points to the same object and pred_as is trakcing.
+                    if (!os1->is_virtual() || !os2->is_virtual()) {
+                      need_update = true;
+                    }
+                  }
 
-                  if (id == pred_as.is_alias(n) && !((pred_os = pred_as.get_object_state(id))->is_virtual())) { // n is escaped.
-                    // materialize 'm' because it has been materialized in save_block.
-                    Node* mv = ensure_object_materialized(m, as, map(), r, block()->init_pnum());
-                    phi->replace_edge(m, mv);
-                    as.escape(id, phi, true);
+                  if (need_update) {
```

Comment:
I don't think we need this `need_update` variable. We can just move this block into line 1907.

Reply:
update it.